sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/implementation/ARMScopeHelper.java

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.experimental.implementation;
+
+import com.azure.core.experimental.http.policy.ARMChallengeAuthenticationPolicy;
+import com.azure.core.http.HttpRequest;
+
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Defines a few utilities. This is a temporary impl class to support {@link ARMChallengeAuthenticationPolicy} until it
+ * moves to azure-resource-manager package.
+ */
+public final class ARMScopeHelper {
+    private ARMScopeHelper() {
+    }
+
+    /**
+     * Generates default scope for oauth2 from the specific request
+     * @param request a http request
+     * @param environment the azure environment with current request
+     * @return the default scope
+     */
+    public static String getDefaultScopeFromRequest(HttpRequest request, AzureEnvironment environment) {
+        return getDefaultScopeFromUrl(request.getUrl().toString().toLowerCase(Locale.ROOT), environment);
+    }
+
+    /**
+     * Generates default scope for oauth2 from the specific request
+     * @param url the url in lower case of a http request
+     * @param environment the azure environment with current request
+     * @return the default scope
+     */
+    static String getDefaultScopeFromUrl(String url, AzureEnvironment environment) {
+        String resource = environment.getManagementEndpoint();
+        for (Map.Entry<String, String> endpoint : environment.getEndpoints().entrySet()) {
+            if (url.contains(endpoint.getValue())) {
+                if (endpoint.getKey().equals(AzureEnvironment.Endpoint.KEYVAULT.identifier())) {
+                    resource = String.format("https://%s/", endpoint.getValue().replaceAll("^\\.*", ""));
```

Comment:
Why does KeyVault use a special format?

Reply:
because its scope is determined via the endpoint as a challenge.